File: backup/cyclesignal_flutter/lib/models/cycle_entry.dart
// Data schema — PRD Section 9. CycleEntry is the only structure persisted.

class MoodEntry {
  final bool present;
  final List<String> types; // low|anxious|irritable|tearful|angry|hopeless
  final int severity; // 1..5
  final bool suddenOnset;

  MoodEntry({
    required this.present,
    required this.types,
    required this.severity,
    required this.suddenOnset,
  });

  Map<String, dynamic> toJson() =>
      {'present': present, 'types': types, 'severity': severity, 'suddenOnset': suddenOnset};

  factory MoodEntry.fromJson(Map<String, dynamic> j) => MoodEntry(
        present: j['present'] ?? false,
        types: (j['types'] as List?)?.map((e) => e.toString()).toList() ?? [],
        severity: (j['severity'] ?? 1) as int,
        suddenOnset: j['suddenOnset'] ?? false,
      );
}

class Severity {
  final bool present;
  final int severity;
  Severity({required this.present, required this.severity});
  Map<String, dynamic> toJson() => {'present': present, 'severity': severity};
  factory Severity.fromJson(Map<String, dynamic> j) =>
      Severity(present: j['present'] ?? false, severity: (j['severity'] ?? 1) as int);
}

class SleepEntry {
  final bool disrupted;
  final int nightsAffected;
  SleepEntry({required this.disrupted, required this.nightsAffected});
  Map<String, dynamic> toJson() => {'disrupted': disrupted, 'nightsAffected': nightsAffected};
  factory SleepEntry.fromJson(Map<String, dynamic> j) =>
      SleepEntry(disrupted: j['disrupted'] ?? false, nightsAffected: (j['nightsAffected'] ?? 0) as int);
}

class PhysicalEntry {
  final Severity? pain;
  final bool acne;
  final bool hairGrowth;
  final Severity? fatigue;
  final SleepEntry? sleep;
  final bool bloating;
  final List<String> other;

  PhysicalEntry({
    this.pain,
    this.acne = false,
    this.hairGrowth = false,
    this.fatigue,
    this.sleep,
    this.bloating = false,
    this.other = const [],
  });

  Map<String, dynamic> toJson() => {
        'pain': pain?.toJson(),
        'acne': acne,
        'hairGrowth': hairGrowth,
        'fatigue': fatigue?.toJson(),
        'sleep': sleep?.toJson(),
        'bloating': bloating,
        'other': other,
      };

  factory PhysicalEntry.fromJson(Map<String, dynamic> j) => PhysicalEntry(
        pain: j['pain'] == null ? null : Severity.fromJson(Map<String, dynamic>.from(j['pain'])),
        acne: j['acne'] ?? false,
        hairGrowth: j['hairGrowth'] ?? false,
        fatigue: j['fatigue'] == null
            ? null
            : Severity.fromJson(Map<String, dynamic>.from(j['fatigue'])),
        sleep: j['sleep'] == null
            ? null
            : SleepEntry.fromJson(Map<String, dynamic>.from(j['sleep'])),
        bloating: j['bloating'] ?? false,
        other: (j['other'] as List?)?.map((e) => e.toString()).toList() ?? [],
      );
}

class BleedingEntry {
  final bool started;
  final bool ended;
  final String? flow; // none|spotting|light|medium|heavy
  BleedingEntry({this.started = false, this.ended = false, this.flow});
  Map<String, dynamic> toJson() => {'started': started, 'ended': ended, 'flow': flow};
  factory BleedingEntry.fromJson(Map<String, dynamic> j) => BleedingEntry(
        started: j['started'] ?? false,
        ended: j['ended'] ?? false,
        flow: j['flow'],
      );
}

class FunctionalImpact {
  final bool present;
  final List<String> types; // missed_work|cancelled_plans|relationship_disruption|unable_to_care_for_self
  final String? severity; // mild|moderate|severe
  FunctionalImpact({this.present = false, this.types = const [], this.severity});
  Map<String, dynamic> toJson() => {'present': present, 'types': types, 'severity': severity};
  factory FunctionalImpact.fromJson(Map<String, dynamic> j) => FunctionalImpact(
        present: j['present'] ?? false,
        types: (j['types'] as List?)?.map((e) => e.toString()).toList() ?? [],
        severity: j['severity'],
      );
}

class EmotionSnapshot {
  final double distress;
  final double sadness;
  final double fear;
  final double calmness;
  EmotionSnapshot(
      {required this.distress, required this.sadness, required this.fear, required this.calmness});
  Map<String, dynamic> toJson() =>
      {'distress': distress, 'sadness': sadness, 'fear': fear, 'calmness': calmness};
  factory EmotionSnapshot.fromJson(Map<String, dynamic> j) => EmotionSnapshot(
        distress: (j['distress'] ?? 0).toDouble(),
        sadness: (j['sadness'] ?? 0).toDouble(),
        fear: (j['fear'] ?? 0).toDouble(),
        calmness: (j['calmness'] ?? 0).toDouble(),
      );
}

class CycleEntry {
  final String id;
  final String date; // ISO yyyy-MM-dd
  final int cycleDay;
  final int cycleNumber;
  final MoodEntry? mood;
  final PhysicalEntry physical;
  final BleedingEntry bleeding;
  final FunctionalImpact functionalImpact;
  final List<String> medications;
  final EmotionSnapshot? emotionSnapshot;
  final String conversationTranscript;
  final String source; // voice|text|synthetic
  final bool safetyFlagged;
  final String consentTimestamp;

  CycleEntry({
    required this.id,
    required this.date,
    required this.cycleDay,
    required this.cycleNumber,
    this.mood,
    required this.physical,
    required this.bleeding,
    required this.functionalImpact,
    this.medications = const [],
    this.emotionSnapshot,
    this.conversationTranscript = '',
    required this.source,
    this.safetyFlagged = false,
    required this.consentTimestamp,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'date': date,
        'cycleDay': cycleDay,
        'cycleNumber': cycleNumber,
        'mood': mood?.toJson(),
        'physical': physical.toJson(),
        'bleeding': bleeding.toJson(),
        'functionalImpact': functionalImpact.toJson(),
        'medications': medications,
        'emotionSnapshot': emotionSnapshot?.toJson(),
        'conversationTranscript': conversationTranscript,
        'source': source,
        'safetyFlagged': safetyFlagged,
        'consentTimestamp': consentTimestamp,
      };

  factory CycleEntry.fromJson(Map<String, dynamic> j) => CycleEntry(
        id: j['id'],
        date: j['date'],
        cycleDay: j['cycleDay'],
        cycleNumber: j['cycleNumber'],
        mood: j['mood'] == null ? null : MoodEntry.fromJson(Map<String, dynamic>.from(j['mood'])),
        physical: PhysicalEntry.fromJson(Map<String, dynamic>.from(j['physical'])),
        bleeding: BleedingEntry.fromJson(Map<String, dynamic>.from(j['bleeding'])),
        functionalImpact:
            FunctionalImpact.fromJson(Map<String, dynamic>.from(j['functionalImpact'])),
        medications: (j['medications'] as List?)?.map((e) => e.toString()).toList() ?? [],
        emotionSnapshot: j['emotionSnapshot'] == null
            ? null
            : EmotionSnapshot.fromJson(Map<String, dynamic>.from(j['emotionSnapshot'])),
        conversationTranscript: j['conversationTranscript'] ?? '',
        source: j['source'] ?? 'synthetic',
        safetyFlagged: j['safetyFlagged'] ?? false,
        consentTimestamp: j['consentTimestamp'] ?? '',
      );
}

File: backup/cyclesignal_rn/server/src/lib/cycleEntry.ts
import { randomUUID } from 'node:crypto';
import { z } from 'zod';

// Symptom-extraction contract — PRD Section 5.2 / Section 9.
// GPT-5.4 Mini (real: gpt-4o-mini) returns this flat, fully-required shape via
// OpenAI Structured Outputs (`strict: true`). Flat primitives keep us inside the
// strict-mode rules (every field required, no nullable objects). The server then
// assembles the full typed `CycleEntry` the client persists.

export const MOOD_TYPES = ['low', 'anxious', 'irritable', 'tearful', 'angry', 'hopeless'] as const;
export const FUNCTIONAL_TYPES = [
  'missed_work',
  'cancelled_plans',
  'relationship_disruption',
  'unable_to_care_for_self',
] as const;
export const FLOW_LEVELS = ['none', 'spotting', 'light', 'medium', 'heavy'] as const;
export const FI_SEVERITY = ['none', 'mild', 'moderate', 'severe'] as const;

// JSON schema handed to OpenAI (strict structured outputs).
export const extractionJsonSchema = {
  type: 'object',
  additionalProperties: false,
  properties: {
    moodPresent: { type: 'boolean' },
    moodTypes: { type: 'array', items: { type: 'string', enum: MOOD_TYPES } },
    moodSeverity: { type: 'integer', description: '0 if no mood symptom, else 1-5' },
    moodSuddenOnset: { type: 'boolean' },
    painPresent: { type: 'boolean' },
    painSeverity: { type: 'integer', description: '0 if no pain, else 1-5' },
    acne: { type: 'boolean' },
    hairGrowth: { type: 'boolean' },
    fatiguePresent: { type: 'boolean' },
    fatigueSeverity: { type: 'integer', description: '0 if no fatigue, else 1-5' },
    sleepDisrupted: { type: 'boolean' },
    sleepNightsAffected: { type: 'integer' },
    bloating: { type: 'boolean' },
    bleedingStarted: { type: 'boolean' },
    bleedingEnded: { type: 'boolean' },
    flow: { type: 'string', enum: FLOW_LEVELS },
    functionalImpactPresent: { type: 'boolean' },
    functionalImpactTypes: { type: 'array', items: { type: 'string', enum: FUNCTIONAL_TYPES } },
    functionalImpactSeverity: { type: 'string', enum: FI_SEVERITY },
    medications: { type: 'array', items: { type: 'string' } },
  },
  required: [
    'moodPresent',
    'moodTypes',
    'moodSeverity',
    'moodSuddenOnset',
    'painPresent',
    'painSeverity',
    'acne',
    'hairGrowth',
    'fatiguePresent',
    'fatigueSeverity',
    'sleepDisrupted',
    'sleepNightsAffected',
    'bloating',
    'bleedingStarted',
    'bleedingEnded',
    'flow',
    'functionalImpactPresent',
    'functionalImpactTypes',
    'functionalImpactSeverity',
    'medications',
  ],
} as const;

export const extractionSchema = z.object({
  moodPresent: z.boolean(),
  moodTypes: z.array(z.enum(MOOD_TYPES)),
  moodSeverity: z.number().int(),
  moodSuddenOnset: z.boolean(),
  painPresent: z.boolean(),
  painSeverity: z.number().int(),
  acne: z.boolean(),
  hairGrowth: z.boolean(),
  fatiguePresent: z.boolean(),
  fatigueSeverity: z.number().int(),
  sleepDisrupted: z.boolean(),
  sleepNightsAffected: z.number().int(),
  bloating: z.boolean(),
  bleedingStarted: z.boolean(),
  bleedingEnded: z.boolean(),
  flow: z.enum(FLOW_LEVELS),
  functionalImpactPresent: z.boolean(),
  functionalImpactTypes: z.array(z.enum(FUNCTIONAL_TYPES)),
  functionalImpactSeverity: z.enum(FI_SEVERITY),
  medications: z.array(z.string()),
});
export type Extraction = z.infer<typeof extractionSchema>;

// Emotion snapshot from EVI 3 prosody — PRD Section 9. Stored, never diagnostic.
export const emotionSnapshotSchema = z
  .object({
    distress: z.number(),
    sadness: z.number(),
    fear: z.number(),
    calmness: z.number(),
  })
  .nullable();

// The context the client supplies alongside the transcript (cycle position etc.).
export const extractRequestSchema = z.object({
  transcript: z.string().max(8000),
  cycleDay: z.number().int().positive(),
  cycleNumber: z.number().int().positive(),
  consentTimestamp: z.string(),
  date: z.string().optional(),
  source: z.enum(['voice', 'text', 'synthetic']).default('voice'),
  emotionSnapshot: emotionSnapshotSchema.default(null),
});
export type ExtractRequest = z.infer<typeof extractRequestSchema>;

function clampSeverity(n: number): 1 | 2 | 3 | 4 | 5 {
  return Math.max(1, Math.min(5, Math.round(n))) as 1 | 2 | 3 | 4 | 5;
}

// Map the flat model output + request context into the full CycleEntry the client
// stores verbatim (PRD Section 9). Mirrors the on-device extractor's output shape
// so the client treats server and on-device results identically.
export function assembleCycleEntry(x: Extraction, ctx: ExtractRequest) {
  return {
    id: randomUUID(),
    date: ctx.date ?? new Date().toISOString().slice(0, 10),
    cycleDay: ctx.cycleDay,
    cycleNumber: ctx.cycleNumber,
    mood: x.moodPresent
      ? {
          present: true,
          types: x.moodTypes.length ? x.moodTypes : ['low'],
          severity: clampSeverity(x.moodSeverity || 3),
          suddenOnset: x.moodSuddenOnset,
        }
      : null,
    physical: {
      pain: x.painPresent ? { present: true, severity: clampSeverity(x.painSeverity || 3) } : null,
      acne: x.acne,
      hairGrowth: x.hairGrowth,
      fatigue: x.fatiguePresent
        ? { present: true, severity: clampSeverity(x.fatigueSeverity || 4) }
        : null,
      sleep: x.sleepDisrupted
        ? { disrupted: true, nightsAffected: Math.max(0, x.sleepNightsAffected) }
        : null,
      bloating: x.bloating,
      other: [] as string[],
    },
    bleeding: {
      started: x.bleedingStarted,
      ended: x.bleedingEnded,
      flow: x.flow === 'none' ? null : x.flow,
    },
    functionalImpact: {
      present: x.functionalImpactPresent,
      types: x.functionalImpactTypes,
      severity: x.functionalImpactSeverity === 'none' ? null : x.functionalImpactSeverity,
    },
    medications: x.medications,
    emotionSnapshot: ctx.emotionSnapshot,
    conversationTranscript: ctx.transcript.slice(0, 2000),
    source: ctx.source,
    safetyFlagged: false,
    consentTimestamp: ctx.consentTimestamp,
  };
}

// The extraction system prompt — PRD 5.2 / 6.1. Non-diagnostic, extract-only.
export const EXTRACTION_SYSTEM = `You extract structured menstrual-cycle symptom data from a short conversation transcript between a health-tracking assistant ("CycleSignal") and a user.

Rules:
- Extract ONLY what the user explicitly states or clearly implies. Do not invent symptoms.
- If a symptom is not mentioned, set its boolean to false and its severity to 0.
- Severities are 1 (mild) to 5 (severe). Use 0 only to mean "not present".
- Never produce a diagnosis or clinical label. You only structure what was said.
- "sudden onset" / "crash" / "out of nowhere" → moodSuddenOnset true.
- Functional impact = missing work, cancelling plans, relationship strain, or being unable to care for oneself.
- medications: only list a medication/supplement if the user named one.`;
